File: flutter_admin_app/lib/services/kubernetes_service.dart
// Service for Kubernetes API calls
import 'package:dio/dio.dart';
import '../models/kubernetes_model.dart';

class KubernetesService {
  final Dio dio;
  final String baseUrl;

  KubernetesService({required this.dio, required this.baseUrl});

  Future<K8sDeployment> createDeployment({
    required String name,
    required String namespace,
    required String imageUrl,
    required int replicas,
  }) async {
    try {
      final response = await dio.post(
        '$baseUrl/api/deployment/kubernetes/deployments',
        queryParameters: {
          'name': name,
          'namespace': namespace,
          'imageUrl': imageUrl,
          'replicas': replicas,
        },
      );
      return K8sDeployment.fromJson(response.data as Map<String, dynamic>);
    } catch (e) {
      rethrow;
    }
  }

  Future<K8sDeployment?> getDeployment(String deploymentId) async {
    try {
      final response = await dio.get('$baseUrl/api/deployment/kubernetes/deployments/$deploymentId');
      return K8sDeployment.fromJson(response.data as Map<String, dynamic>);
    } catch (e) {
      return null;
    }
  }

  Future<List<K8sDeployment>> listDeployments() async {
    try {
      final response = await dio.get('$baseUrl/api/deployment/kubernetes/deployments');
      return (response.data as List)
          .map((e) => K8sDeployment.fromJson(e as Map<String, dynamic>))
          .toList();
    } catch (e) {
      return [];
    }
  }

  Future<void> scaleDeployment(String deploymentId, int replicas) async {
    try {
      await dio.put(
        '$baseUrl/api/deployment/kubernetes/deployments/$deploymentId/scale',
        queryParameters: {'replicas': replicas},
      );
    } catch (e) {
      rethrow;
    }
  }

  Future<void> updateDeploymentImage(String deploymentId, String imageUrl) async {
    try {
      await dio.put(
        '$baseUrl/api/deployment/kubernetes/deployments/$deploymentId/image',
        queryParameters: {'imageUrl': imageUrl},
      );
    } catch (e) {
      rethrow;
    }
  }

  Future<K8sPod> createPod({
    required String podName,
    required String namespace,
    required String deploymentId,
    required String containerImage,
  }) async {
    try {
      final response = await dio.post(
        '$baseUrl/api/deployment/kubernetes/pods',
        queryParameters: {
          'podName': podName,
          'namespace': namespace,
          'deploymentId': deploymentId,
          'containerImage': containerImage,
        },
      );
      return K8sPod.fromJson(response.data as Map<String, dynamic>);
    } catch (e) {
      rethrow;
    }
  }

  Future<K8sPod?> getPod(String podId) async {
    try {
      final response = await dio.get('$baseUrl/api/deployment/kubernetes/pods/$podId');
      return K8sPod.fromJson(response.data as Map<String, dynamic>);
    } catch (e) {
      return null;
    }
  }

  Future<List<K8sPod>> listPodsByDeployment(String deploymentId) async {
    try {
      final response = await dio.get('$baseUrl/api/deployment/kubernetes/pods/deployment/$deploymentId');
      return (response.data as List)
          .map((e) => K8sPod.fromJson(e as Map<String, dynamic>))
          .toList();
    } catch (e) {
      return [];
    }
  }

  Future<void> updatePodStatus(String podId, String status) async {
    try {
      await dio.put(
        '$baseUrl/api/deployment/kubernetes/pods/$podId/status',
        queryParameters: {'status': status},
      );
    } catch (e) {
      rethrow;
    }
  }

  Future<K8sService> createService({
    required String serviceName,
    required String namespace,
    required int port,
    required String selector,
  }) async {
    try {
      final response = await dio.post(
        '$baseUrl/api/deployment/kubernetes/services',
        queryParameters: {
          'serviceName': serviceName,
          'namespace': namespace,
          'port': port,
          'selector': selector,
        },
      );
      return K8sService.fromJson(response.data as Map<String, dynamic>);
    } catch (e) {
      rethrow;
    }
  }

  Future<K8sService?> getService(String serviceId) async {
    try {
      final response = await dio.get('$baseUrl/api/deployment/kubernetes/services/$serviceId');
      return K8sService.fromJson(response.data as Map<String, dynamic>);
    } catch (e) {
      return null;
    }
  }

  Future<List<K8sService>> listServices() async {
    try {
      final response = await dio.get('$baseUrl/api/deployment/kubernetes/services');
      return (response.data as List)
          .map((e) => K8sService.fromJson(e as Map<String, dynamic>))
          .toList();
    } catch (e) {
      return [];
    }
  }

  Future<ClusterHealth?> getClusterHealth() async {
    try {
      final response = await dio.get('$baseUrl/api/deployment/kubernetes/health');
      return ClusterHealth.fromJson(response.data as Map<String, dynamic>);
    } catch (e) {
      return null;
    }
  }
}

File: src/main/java/org/example/service/KubernetesService.java
package org.example.service;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Kubernetes Service
 * Kubernetes deployment and orchestration management
 */
public class KubernetesService {
    
    private final Map<String, K8sDeployment> deployments = new ConcurrentHashMap<>();
    private final Map<String, K8sPod> pods = new ConcurrentHashMap<>();
    private final Map<String, K8sService> services = new ConcurrentHashMap<>();
    
    /**
     * Create Kubernetes deployment
     */
    public K8sDeployment createDeployment(String name, String namespace, String imageUrl, int replicas) {
        K8sDeployment deployment = new K8sDeployment(
                UUID.randomUUID().toString(),
                name,
                namespace,
                imageUrl,
                replicas,
                System.currentTimeMillis()
        );
        deployments.put(deployment.deploymentId, deployment);
        return deployment;
    }
    
    /**
     * Get deployment
     */
    public K8sDeployment getDeployment(String deploymentId) {
        return deployments.get(deploymentId);
    }
    
    /**
     * List all deployments
     */
    public Collection<K8sDeployment> listDeployments() {
        return new ArrayList<>(deployments.values());
    }
    
    /**
     * Scale deployment
     */
    public void scaleDeployment(String deploymentId, int replicas) {
        K8sDeployment deployment = deployments.get(deploymentId);
        if (deployment != null) {
            deployment.desiredReplicas = replicas;
            deployment.lastScaledAt = System.currentTimeMillis();
        }
    }
    
    /**
     * Update deployment image
     */
    public void updateDeploymentImage(String deploymentId, String newImageUrl) {
        K8sDeployment deployment = deployments.get(deploymentId);
        if (deployment != null) {
            deployment.imageUrl = newImageUrl;
            deployment.lastUpdatedAt = System.currentTimeMillis();
        }
    }
    
    /**
     * Create pod
     */
    public K8sPod createPod(String podName, String namespace, String deploymentId, String containerImage) {
        K8sPod pod = new K8sPod(
                UUID.randomUUID().toString(),
                podName,
                namespace,
                deploymentId,
                containerImage,
                System.currentTimeMillis()
        );
        pods.put(pod.podId, pod);
        return pod;
    }
    
    /**
     * Get pod
     */
    public K8sPod getPod(String podId) {
        return pods.get(podId);
    }
    
    /**
     * List pods by deployment
     */
    public List<K8sPod> listPodsByDeployment(String deploymentId) {
        return pods.values().stream()
                .filter(p -> p.deploymentId.equals(deploymentId))
                .toList();
    }
    
    /**
     * Update pod status
     */
    public void updatePodStatus(String podId, String status) {
        K8sPod pod = pods.get(podId);
        if (pod != null) {
            pod.status = status;
            if ("Running".equals(status)) {
                pod.startedAt = System.currentTimeMillis();
            } else if ("Terminated".equals(status)) {
                pod.terminatedAt = System.currentTimeMillis();
            }
        }
    }
    
    /**
     * Create service
     */
    public K8sService createService(String serviceName, String namespace, int port, String selector) {
        K8sService service = new K8sService(
                UUID.randomUUID().toString(),
                serviceName,
                namespace,
                port,
                selector,
                System.currentTimeMillis()
        );
        services.put(service.serviceId, service);
        return service;
    }
    
    /**
     * Get service
     */
    public K8sService getService(String serviceId) {
        return services.get(serviceId);
    }
    
    /**
     * List services
     */
    public Collection<K8sService> listServices() {
        return new ArrayList<>(services.values());
    }
    
    /**
     * Get cluster health
     */
    public Map<String, Object> getClusterHealth() {
        Map<String, Object> health = new HashMap<>();
        
        int totalPods = pods.size();
        long runningPods = pods.values().stream()
                .filter(p -> "Running".equals(p.status))
                .count();
        long pendingPods = pods.values().stream()
                .filter(p -> "Pending".equals(p.status))
                .count();
        long failedPods = pods.values().stream()
                .filter(p -> "Failed".equals(p.status))
                .count();
        
        double healthPercent = totalPods > 0 ? (double) runningPods / totalPods * 100 : 0;
        
        health.put("totalPods", totalPods);
        health.put("runningPods", runningPods);
        health.put("pendingPods", pendingPods);
        health.put("failedPods", failedPods);
        health.put("healthPercent", String.format("%.2f%%", healthPercent));
        health.put("totalDeployments", deployments.size());
        health.put("totalServices", services.size());
        
        return health;
    }
    
    /**
     * K8s Deployment
     */
    public static class K8sDeployment {
        public String deploymentId;
        public String name;
        public String namespace;
        public String imageUrl;
        public int desiredReplicas;
        public int readyReplicas = 0;
        public long createdAt;
        public long lastUpdatedAt = 0;
        public long lastScaledAt = 0;
        
        public K8sDeployment(String deploymentId, String name, String namespace, String imageUrl, int replicas, long createdAt) {
            this.deploymentId = deploymentId;
            this.name = name;
            this.namespace = namespace;
            this.imageUrl = imageUrl;
            this.desiredReplicas = replicas;
            this.createdAt = createdAt;
        }
    }
    
    /**
     * K8s Pod
     */
    public static class K8sPod {
        public String podId;
        public String podName;
        public String namespace;
        public String deploymentId;
        public String containerImage;
        public String status = "Pending";
        public long createdAt;
        public long startedAt = 0;
        public long terminatedAt = 0;
        public int restartCount = 0;
        public List<String> logs = Collections.synchronizedList(new ArrayList<>());
        
        public K8sPod(String podId, String podName, String namespace, String deploymentId, String containerImage, long createdAt) {
            this.podId = podId;
            this.podName = podName;
            this.namespace = namespace;
            this.deploymentId = deploymentId;
            this.containerImage = containerImage;
            this.createdAt = createdAt;
        }
        
        public long getAge() {
            return System.currentTimeMillis() - createdAt;
        }
    }
    
    /**
     * K8s Service
     */
    public static class K8sService {
        public String serviceId;
        public String serviceName;
        public String namespace;
        public int port;
        public String selector;
        public long createdAt;
        public List<String> endpoints = Collections.synchronizedList(new ArrayList<>());
        
        public K8sService(String serviceId, String serviceName, String namespace, int port, String selector, long createdAt) {
            this.serviceId = serviceId;
            this.serviceName = serviceName;
            this.namespace = namespace;
            this.port = port;
            this.selector = selector;
            this.createdAt = createdAt;
        }
    }
}
